hw/ip/flash_ctrl/doc/_index.md

```diff
@@ -65,40 +75,130 @@ The physical controller supports the following features
+The scramble keys are provided from an external static block such as the OTP.
+
+### Flash Scrambling
+
+Flash scrambling is built using the [XEX tweakable block cipher](https://en.wikipedia.org/wiki/Disk_encryption_theory#Xor%E2%80%93encrypt%E2%80%93xor_(XEX)).
+
+When a read transaction is sent to flash, the following steps are taken:
+*  The tweak is calculated using the transaction address and a secret address key through a galois multiplier.
+*  The data content is read out of flash.
+*  If the data content is scrambled, the tweak is XOR'd with the scrambled text and then decrypted through the prince block cipher using a secret data key.
+*  The output of the prince cipher is XOR'd again with the tweak and the final results are presented
+*  If the data content is not scrambled, the prince and XOR steps are skipped and data provided directly back to the requestor.
+
+When a program transaction is sent to flash, the same steps are taken if the address in question has scrambling enabled.
+During a program, the text is encrypted through the prince block cipher.
+
+Note, at the moment scrambling is enabled on the entire 2nd bank of flash, but in the future will be a page boundary attribute.
+
+### Flash Read Pipeline
+
+Since the system host reads directly from the flash for instructions, it is critical to not add significant latency during read, especially if de-scrambling is required.
+As such, the flash read is actually a two stage pipeline, where each stage can take multiple cycles.
+
+Additionally, since the flash word size is typically larger than the bus word, recently read flash entries are locally cached.
+The cache behaves as a highly simplified read-only-cache and holds by default 4 flash words per flash bank.
+
+When a read transaction is sent to flash, the following steps are taken:
+*  A check is performed against the local cache
+   * If there is a hit (either the entry is already in cache, or the entry is currently being processed), the transacton is immediately forwarded to the response queue.
+   * If there is not a hit, an entry in the local cache is selected for allocation (round robin arbitration) and a flash read is issued.
+*  When the flash read completes, its descrambling attributes are checked:
+   * If descrambling is required, the read data begins the descrambling phase - at this time, a new flash read can be issued for the following transaction.
+   * if descrambling is not required, the descrambling phase is skipped and the transaction is pushed to the response queue.
+*  When the descrmbling is complete, the descrambled text is pushed to the response queue.
+
+The following diagram shows how the flash read pipeline timing works.
+![Flash Read Pipeline](flash_read_pipeline.svg)
+
+
+In this example, the first two host requests trigger a full sequence.
+The third host requests immediately hits in the local cache and responds in order after the first two.
+
+### Accessing Information Partition
```

Comment:
I suppose scrambling & ECC also apply to info partitions.

Reply:
yes that's right. I should add that.
This isn't here yet, but eventually when life cycle support comes online, we are going to have a few info pages that are NEVER scrambled (this is for manufacturing purposes), I'll add that point somewhere.

Reply:
Understood.

Reply:
added more text to explain.